keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,360 @@
+
+Here are some API illustations.
+
+##### Node Object extension
+
+```yaml
+type ExtensionFoo struct {
+ TypeMeta
+ ObjectMeta
+ Spec ExtensionFooStatus
+}
+
+type ExtensionFooStatus {
+  v1beta1.LeaseSpec
+  OtherFooStatus interface{}
+}
+
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
+...
+}
+
+type SystemExtension struct {
+   Reference ObjectReference
+   AffectsReadiness bool // Ignore non-critical node extensions from node readiness
+}
+```
+
+##### Sample Extension Status Object
+
+```yaml
+type DevicePluginExtension struct {
+ TypeMeta
+ ObjectMeta
+ Status DevicePluginExtensionStatus
+}
+
+type DevicePluginExtensionStatus struct {
+  LeaseSpec
+  ComputeResource struct{ // embedded for brevity
+    Name: “foo.com/bar”
+    Quantity: x // Sample fields for illustration purposes
+    UnHealthy: y
+    DeviceIDs []string
+  }
+}
+
+```
+
+One of the goals for this proposal is to evaluate the viability and usefullness of combining Status and Health for extensions.
+
+### Approach B - Use Readiness of Node extension pods (Preferred)
+
+This approach assumes that all node level extensions will get deployed as kubernetes pods and that the health of those extension pods can be exposed via existing Readiness probes.
+
+Similar to the previous approach `A`, a special taint will be used to track the overall readiness of the node and the node controller and kubelet will manage the special taint.
+
+The Node Object will be extended to include a list of system extensions that are identified via a label selector and a namespace.
+
+```yaml
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
+...
+}
+
+type SystemExtension struct {
+   Selector LabelSelector
+   Namespace Namespace
+   AffectsReadiness bool // Ignore non-critical node extensions from node readiness
+}
+```
+
+Node selector will take into account `Readiness` all the pods that match the selector(s) specified in the Node Spec `SystemExtensions` field to determine Node Readiness.
+
+The individual system extensions are required to detect kubelet crash (or restart) quickly and reflect that in their Readiness.
+If the Node Controller observes that all the system extension pods are Ready after the kubelet places the special meta readiness taint, it will remove that taint immediately.
+If any of the pods become unready, then the node controller will place the meta readiness taint on the nodes.
+
+This approach has the advantage of not introducing additional API concepts to track state as illustrated in Approach A.
+Side-car containers can be used to expose Readiness more easily than in approach A where it involves updating API objects directly.
+
+Handling of kubelet restarts can be tricky though since there is a chance that the extension pods may not reflect their connections with the kubelet soon after the kubelet restarts.
+
+### Approach C - Use conditions
```

Comment:
Can we also contrast using taints?  As I understand it, Conditions was the plan, then we realized we needed a mechanism to tolerate certain Conditions, and so we introduced taints & tolerations (which are also generally useful).

Reply:
Yes, taints will be used in the current preferred approach.